walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,94 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<Event> eventsSortedByEnd = sortEvents(events, false);
+
+    clearIrrelevantEvents(eventsSortedByEnd, request.getAttendees());
+
+    return findAvailableTimeSlots(eventsSortedByEnd, request.getDuration());
+  }
+
+  private Collection<TimeRange> findAvailableTimeSlots(
+      Collection<Event> eventsSortedByEnd, long minDuration) {
+    Collection<TimeRange> options = new ArrayList<>();
+
+    if (eventsSortedByEnd.isEmpty()) {
+      options.add(TimeRange.WHOLE_DAY);
+    } else {
+      Event event = getFirstEvent(eventsSortedByEnd);
+      addTimeRangeIfPossible(options, TimeRange.START_OF_DAY, event.getWhen().start(), minDuration);
+
+      Iterator<Event> eventsIter = eventsSortedByEnd.iterator();
+      while (eventsIter.hasNext()) {
+        Event nextEvent = eventsIter.next();
+        if (!nextEvent.getWhen().overlaps(event.getWhen())) {
+          addTimeRangeIfPossible(
+              options, event.getWhen().end(), nextEvent.getWhen().start(), minDuration);
+        }
+        event = nextEvent;
+      }
+
+      addTimeRangeIfPossible(options, event.getWhen().end(), TimeRange.END_OF_DAY, minDuration);
+    }
+
+    if (TimeRange.WHOLE_DAY.duration() < minDuration) {
+      options.clear();
+    }
+    return options;
+  }
+
+  private void addTimeRangeIfPossible(
+      Collection<TimeRange> options, int beginTime, int endTime, long minDuration) {
+    TimeRange option = TimeRange.fromStartEnd(beginTime, endTime, endTime == TimeRange.END_OF_DAY);
+    if (option.duration() >= minDuration) {
+      options.add(option);
+    }
+  }
+
+  private Event getFirstEvent(Collection<Event> events) {
+    List<Event> eventsSorted = sortEvents(events, true);
+    return eventsSorted.get(0);
+  }
+
+  private List<Event> sortEvents(Collection<Event> events, boolean sortByStart) {
+    List<Event> eventsSorted = new ArrayList<>();
+    eventsSorted.addAll(events);
+
+    Collections.sort(
+        eventsSorted,
+        new Comparator<Event>() {
+          @Override
```

Comment:
I think it would be better to declare the comparator at the top of the class rather than inline

Reply:
You declare two separate ones as done in TimeRange.java-one for ORDER_BY_START and another for ORDER_BY_END

Reply:
Ah that's a good idea, I'll do that then.